### ISSUE TITLE
failed to start SSH session: unable to exchange encryption keys

### ISSUE THREAD

**Issue by jalil1408** [libgit2/libgit2sharp#1809]
var r = new Repository(working);
            r.Network.Remotes.Add("remote", @"git@gitlab.com:user1234/test.git");

            r.Network.Push(r.Network.Remotes["remote"], @"refs/heads/master", new PushOptions
            {
                CredentialsProvider =
                //(url, user, credentials) => new UsernamePasswordCredentials
                //{
                //    Username = "user1234",
                //    Password = "********"
                //}
                (url, user, credentials) => new SshUserKeyCredentials
                {
                    Username = "git",
                    Passphrase = string.Empty,
                    PublicKey = @"C:\Users\me\.ssh\github.pub",
                    PrivateKey = @"C:\Users\me\.ssh\github"
                }
            });

both UsernamePasswordCredentials and SshUserKeyCredentials give: 

failed to start SSH session: unable to exchange encryption keys

**Comment by bording** [comment]
LibGit2Sharp does not support SSH.

**Comment by jalil1408** [comment]
> 
> 
> LibGit2Sharp does not support SSH.

will it support SSH soon?

**Comment by bording** [comment]
No, the official packages are unlikely to ever support it. We would have to ship SSH libraries as part of our packages, and that's not something we're interested in doing. There's far too many maintenance and security burdens involved if we did that.

If you want to use LibGit2Sharp, you'll need to use HTTPS instead of SSH.

**Comment by ethomson** [comment]
I think that the realistic way to support this is to build the ability for libgit2 to understand how to talk to `/usr/bin/ssh` (or `ssh.exe`).  This would be an opt-in experience for people using the library, obviously.  Depending on libssh2 is probably fine for many people, but I think that most users just want ssh to work like it does with git.  https://github.com/libgit2/libgit2/pull/5507

**Comment by bording** [comment]
If libgit2 did implement that kind of change, then yeah LibGit2Sharp could definitely take advantage of that.

Until something like that happens, though, HTTPS is going to be the only supported protocol.

**Comment by ethomson** [comment]
Strongly agree - you shouldn't (can't, realistically) be the security response team for somebody else's product.

**Comment by michaelkruglos** [comment]
libgit2 libraries for other languages (e.g. nodejs) support SSH. Lacking this feature is a significant limitation.

**Comment by KeyPuncher** [comment]
I haven't dug through the code, but what about defining an interface for SSH operations? Users can implement the interface then hand it over to LibGit2Sharp and have it route all SSH operations through the provided interface instance.
